satellite/satellitedb/overlaycache.go

```diff
@@ -6,122 +6,234 @@ package satellitedb
 import (
 	"context"
 	"database/sql"
-	"errors"
 
-	"storj.io/storj/pkg/utils"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/overlay"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
 	dbx "storj.io/storj/satellite/satellitedb/dbx"
-	"storj.io/storj/storage"
 )
 
+var _ overlay.DB = (*overlaycache)(nil)
+
 type overlaycache struct {
 	db *dbx.DB
 }
 
-func (o *overlaycache) Put(key storage.Key, value storage.Value) error {
-	if key.IsZero() {
-		return storage.ErrEmptyKey.New("")
+// Get looks up the node by nodeID
+func (cache *overlaycache) Get(ctx context.Context, id storj.NodeID) (*pb.Node, error) {
+	if id.IsZero() {
+		return nil, overlay.ErrEmptyNode
 	}
-	ctx := context.Background() // TODO: fix
 
-	tx, err := o.db.Open(ctx)
+	node, err := cache.db.Get_OverlayCacheNode_By_NodeId(ctx,
+		dbx.OverlayCacheNode_NodeId(id.Bytes()),
+	)
+	if err == sql.ErrNoRows {
+		return nil, overlay.ErrNodeNotFound
+	}
 	if err != nil {
-		return Error.Wrap(err)
+		return nil, err
 	}
 
-	_, err = tx.Get_OverlayCacheNode_By_Key(ctx, dbx.OverlayCacheNode_Key(key))
-	if err != nil {
-		_, err = tx.Create_OverlayCacheNode(
-			ctx,
-			dbx.OverlayCacheNode_Key(key),
-			dbx.OverlayCacheNode_Value(value),
-		)
+	return convertOverlayNode(node)
+}
+
+// GetAll looks up nodes based on the ids from the overlay cache
+func (cache *overlaycache) GetAll(ctx context.Context, ids storj.NodeIDList) ([]*pb.Node, error) {
+	infos := make([]*pb.Node, len(ids))
+	for i, id := range ids {
+		// TODO: abort on canceled context
+		info, err := cache.Get(ctx, id)
 		if err != nil {
-			return Error.Wrap(utils.CombineErrors(err, tx.Rollback()))
+			continue
 		}
-	} else {
-		updateFields := dbx.OverlayCacheNode_Update_Fields{}
-		updateFields.Value = dbx.OverlayCacheNode_Value(value)
-		_, err := tx.Update_OverlayCacheNode_By_Key(
-			ctx,
-			dbx.OverlayCacheNode_Key(key),
-			updateFields,
-		)
+		infos[i] = info
+	}
+	return infos, nil
+}
+
+// List lists nodes starting from cursor
+func (cache *overlaycache) List(ctx context.Context, cursor storj.NodeID, limit int) ([]*pb.Node, error) {
+	dbxInfos, err := cache.db.Limited_OverlayCacheNode_By_NodeId_GreaterOrEqual(ctx,
```

Comment:
I think we should cover invalid `limit` with min and max.

Reply:
done

Reply:
also created a separate task for reviewing the limit situation in the future